File: apps/chaoxing_app/lib/models/sync_item.dart
enum SyncItemKind { assignment, exam }

enum SyncDisplayStatus { overdue, today, upcoming, unscheduled }

class SyncItem {
  const SyncItem({
    required this.id,
    required this.kind,
    required this.title,
    required this.url,
    required this.sourceTitle,
    required this.status,
    required this.displayStatus,
    this.sourceSendTime,
    this.startAt,
    this.dueAt,
    this.dueInHours,
    this.courseId,
    this.classId,
    this.workId,
    this.examId,
    this.answerId,
    this.sources = const [],
  });

  final String id;
  final SyncItemKind kind;
  final String title;
  final String url;
  final String sourceTitle;
  final String? sourceSendTime;
  final DateTime? startAt;
  final DateTime? dueAt;
  final String status;
  final SyncDisplayStatus displayStatus;
  final int? dueInHours;
  final String? courseId;
  final String? classId;
  final String? workId;
  final String? examId;
  final String? answerId;
  final List<String> sources;

  bool get isExam => kind == SyncItemKind.exam;

  bool get isOverdue => displayStatus == SyncDisplayStatus.overdue;

  bool get isDueSoon =>
      dueAt != null &&
      dueAt!.isAfter(DateTime.now()) &&
      dueAt!.difference(DateTime.now()).inHours <= 72;

  SyncItem copyWith({
    String? id,
    SyncItemKind? kind,
    String? title,
    String? url,
    String? sourceTitle,
    String? sourceSendTime,
    DateTime? startAt,
    DateTime? dueAt,
    String? status,
    SyncDisplayStatus? displayStatus,
    int? dueInHours,
    String? courseId,
    String? classId,
    String? workId,
    String? examId,
    String? answerId,
    List<String>? sources,
  }) {
    return SyncItem(
      id: id ?? this.id,
      kind: kind ?? this.kind,
      title: title ?? this.title,
      url: url ?? this.url,
      sourceTitle: sourceTitle ?? this.sourceTitle,
      sourceSendTime: sourceSendTime ?? this.sourceSendTime,
      startAt: startAt ?? this.startAt,
      dueAt: dueAt ?? this.dueAt,
      status: status ?? this.status,
      displayStatus: displayStatus ?? this.displayStatus,
      dueInHours: dueInHours ?? this.dueInHours,
      courseId: courseId ?? this.courseId,
      classId: classId ?? this.classId,
      workId: workId ?? this.workId,
      examId: examId ?? this.examId,
      answerId: answerId ?? this.answerId,
      sources: sources ?? this.sources,
    );
  }

  factory SyncItem.fromJson(Map<String, dynamic> json) {
    return SyncItem(
      id: json.readString('id'),
      kind: parseKind(json['kind']),
      title: json.readString('title'),
      url: json.readString('url'),
      sourceTitle: json.readString('sourceTitle'),
      sourceSendTime: json.readNullableString('sourceSendTime'),
      startAt: parseDate(json['startAt']),
      dueAt: parseDate(json['dueAt']),
      status: json.readString('status'),
      displayStatus: parseDisplayStatus(json['displayStatus']),
      dueInHours: json['dueInHours'] is num
          ? (json['dueInHours'] as num).round()
          : null,
      courseId: json.readNullableString('courseId'),
      classId: json.readNullableString('classId'),
      workId: json.readNullableString('workId'),
      examId: json.readNullableString('examId'),
      answerId: json.readNullableString('answerId'),
      sources: json['sources'] is List
          ? (json['sources'] as List)
                .whereType<String>()
                .where((source) => source.isNotEmpty)
                .toList()
          : const [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'kind': kind.name,
      'title': title,
      'url': url,
      'sourceTitle': sourceTitle,
      'sourceSendTime': sourceSendTime,
      'startAt': startAt?.toIso8601String(),
      'dueAt': dueAt?.toIso8601String(),
      'status': status,
      'displayStatus': displayStatus.name,
      'dueInHours': dueInHours,
      'courseId': courseId,
      'classId': classId,
      'workId': workId,
      'examId': examId,
      'answerId': answerId,
      'sources': sources,
    };
  }

  static SyncItemKind parseKind(Object? value) {
    return value == 'exam' ? SyncItemKind.exam : SyncItemKind.assignment;
  }

  static SyncDisplayStatus parseDisplayStatus(Object? value) {
    return SyncDisplayStatus.values.firstWhere(
      (status) => status.name == value,
      orElse: () => SyncDisplayStatus.unscheduled,
    );
  }

  static DateTime? parseDate(Object? value) {
    if (value is! String || value.isEmpty) {
      return null;
    }
    return DateTime.tryParse(value)?.toLocal();
  }
}

extension JsonRead on Map<String, dynamic> {
  String readString(String key) {
    final value = this[key];
    return value is String ? value : '';
  }

  String? readNullableString(String key) {
    final value = this[key];
    return value is String && value.isNotEmpty ? value : null;
  }
}

File: packages/chaoxing-domain/src/sync-item.ts
import { readNullableString, readString } from "./json";

export const SyncItemKind = {
  assignment: "assignment",
  exam: "exam",
} as const;

export type SyncItemKind = (typeof SyncItemKind)[keyof typeof SyncItemKind];

export const SyncDisplayStatus = {
  overdue: "overdue",
  today: "today",
  upcoming: "upcoming",
  unscheduled: "unscheduled",
} as const;

export type SyncDisplayStatus =
  (typeof SyncDisplayStatus)[keyof typeof SyncDisplayStatus];

export type SyncItem = {
  id: string;
  kind: SyncItemKind;
  title: string;
  url: string;
  sourceTitle: string;
  dueAt: Date | null;
  sourceSendTime?: string | null;
  startAt?: Date | null;
  status?: string;
  displayStatus?: SyncDisplayStatus;
  dueInHours?: number | null;
  courseId?: string | null;
  classId?: string | null;
  workId?: string | null;
  examId?: string | null;
  answerId?: string | null;
  sources?: readonly string[];
};

export function parseSyncItemKind(value: unknown): SyncItemKind {
  return value === SyncItemKind.exam
    ? SyncItemKind.exam
    : SyncItemKind.assignment;
}

export function parseDisplayStatus(value: unknown): SyncDisplayStatus {
  return value === SyncDisplayStatus.overdue ||
    value === SyncDisplayStatus.today ||
    value === SyncDisplayStatus.upcoming
    ? value
    : SyncDisplayStatus.unscheduled;
}

export function parseDueAt(value: unknown): Date | null {
  if (typeof value !== "string" || value.length === 0) {
    return null;
  }
  const parsed = new Date(value);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

export function syncItemFromJson(json: Record<string, unknown>): SyncItem {
  const rawSources = json.sources;
  return {
    id: readString(json, "id"),
    kind: parseSyncItemKind(json.kind),
    title: readString(json, "title"),
    url: readString(json, "url"),
    sourceTitle: readString(json, "sourceTitle"),
    sourceSendTime: readNullableString(json, "sourceSendTime"),
    startAt: parseDueAt(json.startAt),
    dueAt: parseDueAt(json.dueAt),
    status: readString(json, "status"),
    displayStatus: parseDisplayStatus(json.displayStatus),
    dueInHours: typeof json.dueInHours === "number" ? Math.round(json.dueInHours) : null,
    courseId: readNullableString(json, "courseId"),
    classId: readNullableString(json, "classId"),
    workId: readNullableString(json, "workId"),
    examId: readNullableString(json, "examId"),
    answerId: readNullableString(json, "answerId"),
    sources: Array.isArray(rawSources)
      ? rawSources.filter(
          (source): source is string =>
            typeof source === "string" && source.length > 0,
        )
      : [],
  };
}

export function syncItemToJson(item: SyncItem): Record<string, unknown> {
  return {
    id: item.id,
    kind: item.kind,
    title: item.title,
    url: item.url,
    sourceTitle: item.sourceTitle,
    sourceSendTime: item.sourceSendTime ?? null,
    startAt: item.startAt?.toISOString() ?? null,
    dueAt: item.dueAt?.toISOString() ?? null,
    status: item.status ?? "",
    displayStatus: item.displayStatus ?? SyncDisplayStatus.unscheduled,
    dueInHours: item.dueInHours ?? null,
    courseId: item.courseId ?? null,
    classId: item.classId ?? null,
    workId: item.workId ?? null,
    examId: item.examId ?? null,
    answerId: item.answerId ?? null,
    sources: [...(item.sources ?? [])],
  };
}

export function copySyncItem(
  item: SyncItem,
  patch: Partial<SyncItem>,
): SyncItem {
  return { ...item, ...patch };
}
